File: Python/Symmetric Tree.py
"""
Given a binary tree, check whether it is a mirror of itself (ie, symmetric around its center).
For example, this binary tree is symmetric:
    1
   / \
  2   2
 / \ / \
3  4 4  3
But the following is not:
    1
   / \
  2   2
   \   \
   3    3
"""

# Definition for a  binary tree node
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    # @param root, a tree node
    # @return a boolean
    def isSymmetric(self, root):
        if root is None:
            return True
        if root.left is None and root.right is None and root:
            return True
        if root.right and root.left:
            if root.right.val != root.left.val:
                return False
        else:
            return False
        forward = [root.left]
        backward = [root.right]
        while len(forward) > 0:
            forward_new =[]
            backward_new = []
            while len(forward) >0:
                node1 = forward.pop()
                node2 = backward.pop()
                
                # case 1: all have left subtree and right subtree
                if node1.left and node1.right and node2.left and node2.right:
                    if node1.left.val == node2.right.val and node1.right.val == node2.left.val:
                        forward_new.append(node1.left)
                        forward_new.append(node1.right)
                        backward_new.append(node2.right)
                        backward_new.append(node2.left)
                    else:
                        return False
                # case 2: one has left and one has right
                elif node1.left and node2.right and node1.right is None and node2.left is None:
                    if node1.left.val != node2.right.val:
                        return False
                    forward_new.append(node1.left)
                    backward_new.append(node2.right)
                elif node2.left and node1.right and node1.left is None and node2.right is None:
                    if node2.left.val != node1.right.val:
                        return False
                    forward_new.append(node1.right)
                    backward_new.append(node2.left)
                # case 3: no subtree
                elif node1.left is None and node1.right is None and node2.left is None and node2.right is None:
                    return True
                else:
                    return False
            forward = forward_new
            backward = backward_new
        return True

class Solution:
    # @param root, a tree node
    # @return a boolean
    def isSymmetric(self, root):
        if root is None:
            return True
        return self.isSymmetricHelper(root.left,root.right)
        
    def isSymmetricHelper(self,left,right):
        if left is None and right is None:
            return True
        elif left and right:
            if left.val == right.val:
                return self.isSymmetricHelper(left.left,right.right) and self.isSymmetricHelper(left.right,right.left)
            else:
                return False
        else:
            return False

File: Java/Symmetric Tree.java
/*
Given a binary tree, check whether it is a mirror of itself (ie, symmetric around its center).

For example, this binary tree is symmetric:

    1
   / \
  2   2
 / \ / \
3  4 4  3
But the following is not:
    1
   / \
  2   2
   \   \
   3    3
Note:
Bonus points if you could solve it both recursively and iteratively.

confused what "{1,#,2,3}" means? > read more on how binary tree is serialized on OJ.
*/

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public boolean isSymmetric(TreeNode root) {
        return root == null ? true : isSymmetric(root.left, root.right);
    }
    
    private boolean isSymmetric(TreeNode left, TreeNode right){
        if(left == null && right == null) return true;
        if(left == null || right == null) return false;
        
        return left.val == right.val && isSymmetric(left.left, right.right) && isSymmetric(left.right, right.left);
    }
}

public class Solution {
    public boolean isSymmetric(TreeNode root) {
        if(root == null) return true;
        Stack<TreeNode> stack = new Stack<>();
        stack.push(root.left);
        stack.push(root.right);
        
        while(!stack.empty()){
            TreeNode p = stack.peek();
            stack.pop();
            TreeNode q = stack.peek();
            stack.pop();
            
            if(p == null && q == null) continue;
            if(p == null || q == null) return false;
            if(p.val != q.val) return false;
            
            stack.push(p.left);
            stack.push(q.right);
            
            stack.push(p.right);
            stack.push(q.left);
        }
        
        return true;
    }
}